drivers/ethernet/eth_virtio_net.c

```diff
@@ -0,0 +1,259 @@
+	const struct virtnet_config *config = data->dev->config;
+	struct virtq *vq = virtio_get_virtqueue(config->vdev, VIRTQ_RX(1));
+
+	len -= sizeof(struct _virtio_net_hdr);
+	struct net_pkt *pkt =
+		net_pkt_rx_alloc_with_buffer(data->iface, len, AF_UNSPEC, 0, K_FOREVER);
+
+	if (pkt == NULL) {
+		LOG_ERR("received packet, but could not pass it to the operating system");
+	} else if (net_pkt_write(pkt, &(data->rxb[buf_no][sizeof(struct _virtio_net_hdr)]), len)) {
+		LOG_ERR("could not copy entire received packet");
+		net_pkt_unref(pkt);
+	} else if (net_recv_data(data->iface, pkt)) {
+		LOG_ERR("operating system failed to receive packet");
+		net_pkt_unref(pkt);
+	} else {
+		/* Packet received correctly, no error */
+	}
+	struct virtq_buf vqbuf[] = {{.addr = &(data->rxb[buf_no]), .len = VIRTIO_NET_BUFLEN}};
+
+	virtq_add_buffer_chain(vq, vqbuf, 1, 0, virtnet_rx_cb, priv, K_FOREVER);
+	virtio_notify_virtqueue(config->vdev, VIRTQ_RX(1));
+}
+
+static void virtnet_if_init(struct net_if *iface)
+{
+	ethernet_init(iface);
+	const struct device *dev = net_if_get_device(iface);
+	struct virtnet_data *data = dev->data;
+	const struct virtnet_config *config = dev->config;
+
+	if (dev == NULL) {
+		LOG_ERR("could not access device structure!");
+		return;
+	}
+	data->iface = iface;
+	net_if_set_link_addr(iface, data->mac, sizeof(data->virtio_devcfg->mac), NET_LINK_ETHERNET);
+	struct virtq *vq = virtio_get_virtqueue(config->vdev, VIRTQ_RX(1));
+
+	for (int i = 0; i < CONFIG_ETH_VIRTIO_NET_RX_BUFFERS; i++) {
+		data->rx_cb_data[i].data = data;
+		data->rx_cb_data[i].buf_no = i;
+
+		struct virtq_buf vqbuf[] = {{.addr = &(data->rxb[i]), .len = VIRTIO_NET_BUFLEN}};
+
+		virtq_add_buffer_chain(vq, vqbuf, 1, 0, virtnet_rx_cb, &(data->rx_cb_data[i]),
+				       K_FOREVER);
+		virtio_notify_virtqueue(config->vdev, VIRTQ_RX(1));
+	}
+	LOG_DBG("initialization finished");
+}
+
+static int virtnet_dev_init(const struct device *dev)
+{
+	const struct virtnet_config *config = dev->config;
+	struct virtnet_data *data = dev->data;
+
+	if (config->random_mac) {
+		sys_rand_get(data->mac, sizeof(data->mac));
+		/* make it a locally administered, unicast address */
+		/* (2nd hex digit of 1st byte is 2) */
+		data->mac[0] &= ~(BIT(0) | BIT(2) | BIT(3));
+		data->mac[0] |= BIT(1);
+	}
+
+	data->virtio_devcfg = virtio_get_device_specific_config(config->vdev);
+	if (data->virtio_devcfg == NULL) {
+		LOG_ERR("could not get config struct");
+	}
+	if (virtio_commit_feature_bits(config->vdev)) {
+		LOG_ERR("could not commit feature bits");
+	}
+	LOG_DBG("MAC address is %02x:%02x:%02x:%02x:%02x:%02x", data->mac[0], data->mac[1],
+		data->mac[2], data->mac[3], data->mac[4], data->mac[5]);
+
+	virtio_init_virtqueues(config->vdev, 2, virtnet_enum_queues_cb, NULL);
+	virtio_finalize_init(config->vdev);
+
+	return 0;
+}
+
+static struct ethernet_api virtnet_api = {.iface_api.init = virtnet_if_init,
+					  .get_capabilities = virtnet_get_capabilities,
+					  .send = virtnet_send};
```

Comment:
Prefer using a trailing comma so fields are formatted after a single tab character

```suggestion
static struct ethernet_api virtnet_api = {
	.iface_api.init = virtnet_if_init,
	.get_capabilities = virtnet_get_capabilities,
	.send = virtnet_send,
};
```